File: internal/engine/executor/platform/modules/time.go
package modules

import (
	"time"

	lua "github.com/yuin/gopher-lua"
)

// TimeModule provides functions that Lua scripts can call
type TimeModule struct {
}

// NewTimeModule creates a new TimeModule
func NewTimeModule() *TimeModule {
	return &TimeModule{}
}

// Name returns the name of the module
func (s *TimeModule) Name() string {
	return "time"
}

// GetCurrentTime accepts the optional Go-sryle formatter, and returns the current time string
// If no formatter is provided, the default Go time format is used
func (s *TimeModule) GetCurrentTime(L *lua.LState) int {
	format := L.ToString(1)
	now := time.Now()
	if format == "" {
		L.Push(lua.LString(now.String()))
	} else {
		L.Push(lua.LString(now.Format(format)))
	}
	return 1
}

// Loader loads the Time module into the Lua state
func (s *TimeModule) Loader(L *lua.LState) int {
	mod := L.SetFuncs(L.NewTable(), map[string]lua.LGFunction{
		"now": s.GetCurrentTime,
	})

	fields := map[string]lua.LValue{
		"ANSIC":       lua.LString(time.ANSIC),
		"UNIX":        lua.LString(time.UnixDate),
		"RubyDate":    lua.LString(time.RubyDate),
		"RFC822":      lua.LString(time.RFC822),
		"RFC822Z":     lua.LString(time.RFC822Z),
		"RFC850":      lua.LString(time.RFC850),
		"RFC1123":     lua.LString(time.RFC1123),
		"RFC1123Z":    lua.LString(time.RFC1123Z),
		"RFC3339":     lua.LString(time.RFC3339),
		"RFC3339Nano": lua.LString(time.RFC3339Nano),
		"Kitchen":     lua.LString(time.Kitchen),
		"Stamp":       lua.LString(time.Stamp),
		"StampMilli":  lua.LString(time.StampMilli),
		"StampMicro":  lua.LString(time.StampMicro),
		"StampNano":   lua.LString(time.StampNano),
		"DateTime":    lua.LString(time.DateTime),
		"DateOnly":    lua.LString(time.DateOnly),
		"TimeOnly":    lua.LString(time.TimeOnly),
	}

	for name, field := range fields {
		L.SetField(mod, name, field)
	}

	L.Push(mod)
	return 1
}

File: lua/bindings/time.lua
---@meta

---@module 'time' Time module that provides functions to get the current time
local time

--- Time formats.
---@enum time.Format
local Format = {
  ANSIC = 'Mon Jan _2 15:04:05 2006',
  UnixDate = 'Mon Jan _2 15:04:05 MST 2006',
  RubyDate = 'Mon Jan 02 15:04:05 -0700 2006',
  RFC822 = '02-Jan-06 15:04 MST',
  RFC822Z = '02-Jan-06 15:04 -0700',
  RFC850 = 'Monday, 02-Jan-06 15:04:05 MST',
  RFC1123 = 'Mon, 02 Jan 2006 15:04:05 MST',
  RFC1123Z = 'Mon, 02 Jan 2006 15:04:05 -0700',
  RFC3339 = '2006-01-02T15:04:05Z07:00',
  RFC3339Nano = '2006-01-02T15:04:05.999999999Z07:00',
  Kitchen = '3:04PM',
  Stamp = 'Jan _2 15:04:05',
  StampMilli = 'Jan _2 15:04:05.000',
  StampMicro = 'Jan _2 15:04:05.000000',
  StampNano = 'Jan _2 15:04:05.000000000',
  DateTime = '2006-01-02 15:04:05',
  DateOnly = '2006-01-02',
  TimeOnly = '15:04:05',
}

--- Exported Time formats
---@type time.Format
time.ANSIC = Format.ANSIC
---@type time.Format
time.UnixDate = Format.UnixDate
---@type time.Format
time.RubyDate = Format.RubyDate
---@type time.Format
time.RFC822 = Format.RFC822
---@type time.Format
time.RFC822Z = Format.RFC822Z
---@type time.Format
time.RFC850 = Format.RFC850
---@type time.Format
time.RFC1123 = Format.RFC1123
---@type time.Format
time.RFC1123Z = Format.RFC1123Z
---@type time.Format
time.RFC3339 = Format.RFC3339
---@type time.Format
time.RFC3339Nano = Format.RFC3339Nano
---@type time.Format
time.Kitchen = Format.Kitchen
---@type time.Format
time.Stamp = Format.Stamp
---@type time.Format
time.StampMilli = Format.StampMilli
---@type time.Format
time.StampMicro = Format.StampMicro
---@type time.Format
time.StampNano = Format.StampNano
---@type time.Format
time.DateTime = Format.DateTime
---@type time.Format
time.DateOnly = Format.DateOnly
---@type time.Format
time.TimeOnly = Format.TimeOnly

--- GetCurrentTime returns the current time as a string
---
--- Example:
--- ```
--- local time = require 'time'
---
--- local now = time.now(time.ANSIC)
--- assert(now == 'Mon Jan _2 15:04:05 2006')
---
--- local now = time.now(time.DateTime)
--- assert(now == '2023-01-01 15:04:05')
---
--- now = time.now(time.DateOnly)
--- assert(now == '2023-01-01')
---
--- now = time.now(time.TimeOnly)
--- assert(now == '15:04:05')
---
--- now = time.now()
--- assert(now == '2023-01-01 15:04:05.000000000')
--- ```
--- The default format is the Go time format, which is "2006-01-02 15:04:05.999999999"
---
---@param format time.Format? Formatter to use for the current time
---@return string? time Current time as a string
function time.now(format) end

return time
